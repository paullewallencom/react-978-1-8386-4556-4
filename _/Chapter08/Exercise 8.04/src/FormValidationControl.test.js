import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import FormValidationControl from "./FormValidationControl";
import { act } from "react-dom/test-utils";

test("renders correctly", () => {
  const { container } = render(<FormValidationControl />);

  expect(container).toMatchSnapshot();
});

test("loads and displays form", async () => {
  //Arrange
  const { container } = render(<FormValidationControl />);
  const email = container.querySelector('input[type="text"]');
  const password = container.querySelector('input[type="password"]');

  //Act
  await wait(() => {
    fireEvent.change(email, {
      target: {
        value: "example@email.com",
      },
    });

    fireEvent.change(password, {
      target: {
        value: "password",
      },
    });
  });

  //Assert
  expect(email.value).toBe("example@email.com");
  expect(password.value).toBe("password");
});

test("displays an error message", async () => {
  //Arrange
  const { container, queryByText } = render(<FormValidationControl />);
  const email = container.querySelector('input[type="text"]');
  const password = container.querySelector('input[type="password"]');

  //Act
  await wait(() => {
    fireEvent.change(email, {
      target: {
        value: "",
      },
    });
    fireEvent.change(password, {
      target: {
        value: "",
      },
    });
    fireEvent.blur(email);
    fireEvent.blur(password);
  });

  const emailErrorMessage = queryByText(/Name is Required/);
  const passwordErrorMessage = queryByText(/Min length of Password is 8 chars/);

  expect(emailErrorMessage).not.toBeNull();
  expect(passwordErrorMessage).not.toBeNull();
});

test("displays an error message", async () => {
  // mock info, used in submission, so we can spy on it
  let originalInfo = console.info;

  global.console = { ...global.console, info: jest.fn() };

  // mock setTimeout in submission, so we can advance it
  jest.useFakeTimers();

  //Arrange
  const { container, queryByText } = render(<FormValidationControl />);

  const email = container.querySelector('input[type="text"]');
  const password = container.querySelector('input[type="password"]');
  const form = container.querySelector("form");

  // values that passes validation
  const emailValue = "hey@ho.com";
  const passwordValue = "passwordpass";

  //Act
  await wait(() => {
    fireEvent.change(email, {
      target: {
        value: emailValue,
      },
    });
    fireEvent.change(password, {
      target: {
        value: passwordValue,
      },
    });
    fireEvent.blur(email);
    fireEvent.blur(password);
  });

  let emailErrorMessage = queryByText(/Name is Required/);
  let passwordErrorMessage = queryByText(/Min length of Password is 8 chars/);

  // error messages should not be there if the fields are passed validation
  expect(emailErrorMessage).toBeNull();
  expect(passwordErrorMessage).toBeNull();

  // submit form
  await wait(() => {
    fireEvent.submit(form);
  });

  // advance setTimeout
  act(() => {
    jest.advanceTimersByTime(3000);
  });

  const expectedLog = JSON.stringify(
    { name: emailValue, password: passwordValue },
    null,
    2
  );

  // assert against console.info
  expect(console.info).toHaveBeenCalledWith(expectedLog);

  global.console = { ...global.console, info: originalInfo };
});
