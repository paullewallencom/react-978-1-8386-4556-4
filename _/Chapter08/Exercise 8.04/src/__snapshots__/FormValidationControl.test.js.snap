// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form>
    <label>
      Name*:
      <input
        name="name"
        type="text"
        value=""
      />
    </label>
    <label>
      Password*:
      <input
        name="password"
        type="password"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Login"
    />
  </form>
</div>
`;
